### PR TITLE
refactors update document

### PR DESCRIPTION
**Pull request by Glumli**

### CODE REVIEW

File: src/services/documentService.js [hunk 1]
```diff
@@ -91,58 +104,44 @@ const documentService = {
...
-            documentRoutes.downloadRecord(ownerId, hcDocument.id);
-
-        const encryptedFilesPromise = recordPromise
-            .then(record => createCryptoService(ownerId).encryptBlobs(
-                newAttachments.map(attachment => attachment.file),
-                record.attachment_key,
-            ));
+        const encryptedFilesPromise = attachmentKeyPromise
+            .then(ak =>
+                createCryptoService(ownerId).encryptBlobs(
```
Comment (ibihim) [inline src/services/documentService.js]: I don't know if you like challenges or at how much time pressure you are, but this could be a nice memoization application :smile: 
Reason: spinning up a cryptoService on every updateDocument call might be expensive...

File: src/services/documentService.js [hunk 2]
```diff
@@ -91,58 +104,44 @@ const documentService = {
...
-                    }));
-                return uploadFilePromise;
-            })
-            .then(() => {
-                const customTags = hcDocument.annotations
-                    ? taggingUtils.generateCustomTags(hcDocument.annotations)
-                    : [];
+            .then(([[encryptedFiles, attachmentKey], uploadInformation]) => {
+                const uploadFilesPromise = encryptedFiles.map((encryptedFile, index) => {
+                    newAttachments[index].id = uploadInformation[index].id;
```
Comment (ibihim) [inline src/services/documentService.js]: Can we guarantee order (uploadInformation <=> encryptedFiles)? Otherwise it will cause some very hard to reason issues.

File: src/services/documentService.js [hunk 3]
```diff
@@ -7,6 +7,19 @@ import ValidationError from '../lib/errors/ValidationError';
...
+    const newAttachments = [];
+    const oldAttachments = [];
+    collection.forEach((element) => {
+        if (element[attribute]) {
+            oldAttachments.push(element);
+        } else {
+            newAttachments.push(element);
+        }
+    });
+    return { oldAttachments, newAttachments };
```
[the remark this answers is not in the document]
Comment (ibihim) [reply src/services/documentService.js]: Hint: When trying to solve a problem, try to abtract what you are doing. If you would do so, you could see, that you are solving here a general problem and not a particular one :smile:

File: src/services/documentService.js [hunk 4]
```diff
@@ -91,58 +104,44 @@ const documentService = {
...
                     ownerId,
                     hcDocument.id,
                     hcDocumentUtils.toFhirObject(hcDocument, this.clientId),
-                    customTags,
+                    taggingUtils.generateCustomTags(hcDocument.annotations),
                     attachmentKey,
                 );
+                return Promise.all([
+                    updateFhirRecordPromise,
+                    Promise.all(uploadFilesPromise),
```
Comment (ibihim) [inline src/services/documentService.js]: Is this some common pattern to have Promise.all wrapped in another Promise.all? why???

File: src/services/fhirService.js [hunk 5]
```diff
@@ -6,25 +6,20 @@ import dateUtils from '../lib/dateUtils';
 import hcCrypto from '../lib/crypto';
 import createCryptoService from './cryptoService';
 
-
 const fhirService = {
-    updateFhirRecord(ownerId, recordId, fhirObject, customTags = [], attachmentKey = null) {
+    updateFhirRecord(ownerId, recordId, fhirResource, customTags = [], attachmentKey) {
```
Comment (ibihim) [inline src/services/fhirService.js]: why did you remove attachmenKey = null?

File: src/services/documentService.js [hunk 6]
```diff
@@ -91,58 +104,44 @@ const documentService = {
...
+            .then(ak =>
+                createCryptoService(ownerId).encryptBlobs(
+                    newAttachments.map(attachment => attachment.file),
+                    ak,
+                ));
 
         return Promise.all([
             encryptedFilesPromise,
             newAttachments.length ?
                 documentRoutes.getFileUploadUrls(ownerId, hcDocument.id, newAttachments.length) :
```
Comment (ibihim) [inline src/services/documentService.js]: getFileUploadURLs :smile:

File: src/services/documentService.js (hunk 2 above)
Comment (ibihim) [reply src/services/documentService.js]: I see, it is not important :+1:

File: src/services/documentService.js [hunk 7]
```diff
@@ -91,58 +104,44 @@ const documentService = {
...
+            .then(([[encryptedFiles, attachmentKey], uploadInformation]) => {
+                const uploadFilesPromise = encryptedFiles.map((encryptedFile, index) => {
+                    newAttachments[index].id = uploadInformation[index].id;
+                    return fileRoutes.uploadFile(
+                        uploadInformation[index].sas_token,
+                        encryptedFile);
+                });
                 hcDocument.attachments = [...oldAttachments, ...newAttachments];
-                return this.fhirService.updateFhirRecord(
+                const updateFhirRecordPromise = this.fhirService.updateFhirRecord(
```
Comment (ibihim) [inline src/services/documentService.js]: Why is it `this.fhirService` and not fhirSerivce???

Comment (ibihim) [reply src/services/documentService.js]: updateFHIRRecord....

File: src/services/documentService.js (hunk 4 above)
Comment (Glumli) [reply src/services/documentService.js]: I answered this in Simranns PR already ;)

File: src/services/fhirService.js (hunk 5 above)
Comment (Glumli) [reply src/services/fhirService.js]: Does it make sense to have it? Is null so much better than undefined?

File: src/services/documentService.js (hunk 4 above)
Comment (ibihim) [reply src/services/documentService.js]: And I re-answered to this in Simran's PR too

File: src/services/fhirService.js (hunk 5 above)
Comment (ibihim) [reply src/services/fhirService.js]: So, this indicates, that the value is optional. Just looks strange, that an optional value is followed by a non-optional one. 
I think this is common in TypeScript or so (declaring optionality by that).

File: src/services/documentService.js (hunk 4 above)
Comment (ibihim) [reply src/services/documentService.js]: Ok, got it.

File: src/services/documentService.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Glumli) [reply src/services/documentService.js]: Ah fuck... I was pulling my changes over to a new branch, as the other one had flow. 
But actualy I pulled the method out and also gave it a general name and the general parameter because I did exactly this. Forgot renaming the arrays to sth general after rewrite.

File: src/services/documentService.js (hunk 1 above)
Comment (Glumli) [reply src/services/documentService.js]: We really have to think about how to do this in the future... I liked creating a new one each time, as the commonKey might change :/ And without memoization the propability for using an olc commonKey is smaller

File: src/routes/documentRoutes.js [hunk 8]
```diff
@@ -5,11 +5,11 @@ const apiUrl = config.api;
 
 const documentRoutes = {
 
-    getFileDownloadUrl(ownerId, recordId, fileId) {
+    getFileDownloadURL(ownerId, recordId, fileId) {
         return hcRequest.submit('GET', `${apiUrl}/users/${ownerId}/documents/${recordId}/files/${fileId}/download_access_token`, { authorize: true, ownerId });
     },
 
-    getFileUploadUrls(ownerId, recordId, fileNumber) {
+    getFileUploadURLs(ownerId, recordId, fileNumber) {
```
Comment (ibihim) [inline src/routes/documentRoutes.js]: great

File: src/services/documentService.js (hunk 1 above)
Comment (ibihim) [reply src/services/documentService.js]: ok
